File: 백준/Gold/1092. 배/배.java
import java.util.*;

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();

        List<Integer> cranes = new ArrayList<>();
        List<Integer> boxes = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            cranes.add(sc.nextInt());
        }

        int m = sc.nextInt();

        for (int i = 0; i < m; i++) {
            boxes.add(sc.nextInt());
        }

        cranes.sort(Collections.reverseOrder());
        boxes.sort(Collections.reverseOrder());

        if (cranes.get(0) < boxes.get(0)) {
            System.out.println(-1);
            return;
        }
        
        boolean[] visited = new boolean[m];

        int answer = 0;
        int moved = 0;
        while (moved < m) {
            int idx = 0;

            for (int i = 0; i < n; i++) {
                while (idx < m) {
                    if (!visited[idx] && cranes.get(i) >= boxes.get(idx)) {
                        visited[idx] = true;
                        moved++;
                        idx++;
                        break;
                    }
                    idx++;
                }
            }
            answer++;
        }
        System.out.println(answer);
    }
}

File: 백준/Gold/1092. 배/배.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int n, m;
    vector<int> cranes;
    vector<int> boxes;

    cin >> n;
    for (int i = 0; i < n; i++) {
        int crane_weight;
        cin >> crane_weight;

        cranes.push_back(crane_weight);
    }

    cin >> m;
    for (int i = 0; i < m; i++) {
        int box_weight;
        cin >> box_weight;

        boxes.push_back(box_weight);
    }

    sort(cranes.begin(), cranes.end(), greater<int>());
    sort(boxes.begin(), boxes.end(), greater<int>());

    if (cranes[0] < boxes[0]) {
        cout << -1;
        return 0;
    }

    int answer = 0;
    while (!boxes.empty()) {
        int j = 0;
        for (int i = 0; i < n; i++) {
            if (boxes.size() == 0 && i != 0) {
                answer++;
                cout << answer;
                return 0;
            }
            for (int j = 0; j < boxes.size(); j++) {
                if (cranes[i] >= boxes[j]) {
                    boxes.erase(boxes.begin() + j);
                    break;
                } else {
                    j++;
                }
            }
        }
        answer++;
    }
    
    cout << answer;
    return 0;
}